pkg/api/cached.go

```diff
@@ -140,39 +156,81 @@ func (c *Cached) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolu
 	}
 
 	targetPath := req.GetTargetPath()
-	volumeID := req.GetVolumeId()
-	volumeAttributes := req.GetVolumeContext()
+	// The parent of the target path and can store CSI metadata, it's the name given to the volume mount in the pod spec
+	volumePath := path.Join(targetPath, "..")
+	var version int64
 
-	var cachePath string
-	var targetPermissions os.FileMode
+	// Perform an overlay mount
+	upperdir := path.Join(volumePath, UPPER_DIR)
+	err := os.MkdirAll(upperdir, 0777)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create overlay upper directory %s: %v", upperdir, err)
+	}
 
-	if suffix, exists := volumeAttributes["placeCacheAtPath"]; exists {
-		// running in suffix mode, desired outcome:
-		//  - the mount point is writable by the pod
-		//  - the cache is mounted at the suffix, and is not writable
-		cachePath = path.Join(targetPath, suffix)
-		targetPermissions = 0777
-	} else {
-		// running in unsuffixed mode, desired outcome:
-		//  - the mount point *is* the cache, and is not writable by the pod
-		cachePath = targetPath
-		targetPermissions = 0755
+	upperInfo, err := os.Stat(upperdir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to stat overlay upper directory %s: %v", upperdir, err)
+	}
+	if upperInfo.Mode()&os.ModePerm != 0777 {
+		err = os.Chmod(upperdir, 0777)
+		if err != nil {
+			return nil, fmt.Errorf("failed to change permissions of overlay upper directory %s: %v", upperdir, err)
+		}
 	}
 
-	if err := os.MkdirAll(targetPath, targetPermissions); err != nil {
-		return nil, fmt.Errorf("failed to create target directory %s: %s", targetPath, err)
+	workdir := path.Join(volumePath, WORK_DIR)
+	err = os.MkdirAll(workdir, 0777)
```

Comment:
This did not seem to set the dir to `0777` so I had to add the explicit `chmod` below.